### ISSUE TITLE
Incorrect type inference in dict with literal keys

### ISSUE THREAD

**Issue by Fulguritude** [python/mypy#12577]
**Bug Report**

I have some custom dictionary types where keys are string literals, and a utility which accepts any `str` as a key. Running mypy over this utility returns an "incompatible type" error, when the types are most definitely compatible.

**To Reproduce**

```py
from typing import Dict, Literal, TypeVar

GenericDictValue = TypeVar("GenericDictValue")

def merge_dicts(
    dict1 : Dict[str, GenericDictValue],
    dict2 : Dict[str, GenericDictValue],
    clone : bool                        = True,
) -> Dict[str, GenericDictValue]:
    result = dict1.copy() if clone else dict1
    result.update(dict2)
    return result

CustomDictType = Dict[
    Literal["Alice", "Bob"],
    int,
]
dict1 : CustomDictType = {"Alice" : 15}
dict2 : CustomDictType = {"Bob"   : 12}

merge_dicts(dict1, dict2)
```

**Expected Behavior**

```
Success: no issues found in 1 source file
```

**Actual Behavior**

```
src/playground.py:21: error: Argument 1 to "merge_dicts" has incompatible type "Dict[Literal['Alice', 'Bob'], int]"; expected "Dict[str, int]"
src/playground.py:21: error: Argument 2 to "merge_dicts" has incompatible type "Dict[Literal['Alice', 'Bob'], int]"; expected "Dict[str, int]"
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: mypy 0.941
- Mypy command-line flags: (N/A)
- Mypy configuration options from `mypy.ini` (and other config files): (N/A)
- Python version used: Python 3.10
- Operating system and version: Linux, Pop OS! 21.10

**Additional notes**
Of course, a quick fix (and improvement!) on my end is to replace the `str` key types with a `GenericDictKey` type, but I thought that you should be made aware of this, since it's most definitely a bug (unless python just handles literals in a very bad way, with them not being an actual subtype of a common supertype, which doesn't seem to be the case otherwise).

**Comment by JelleZijlstra** [comment]
Dict is invariant in its key type, so this is correct: https://mypy.readthedocs.io/en/stable/common_issues.html#invariance-vs-covariance
